recirq/quantum_chess/quantum_board.py

```diff
@@ -346,84 +345,75 @@ def sample(self, num_samples: int) -> List[int]:
             rtn = rtn + samples
         return rtn[:num_samples]
 
-    def _set_full_empty_squares_from_probability(self) -> None:
-        self.full_squares = 0
-        self.empty_squares = 0
-        for i, p in enumerate(self.probabilities):
+    def _make_probability_history(self, repetitions, probabilities):
+        probabilities = tuple(probabilities)
+        full_squares = 0
+        empty_squares = 0
+        for i, p in enumerate(probabilities):
             if p == 1:
-                self.full_squares = set_nth_bit(i, self.full_squares, True)
+                full_squares = set_nth_bit(i, full_squares, True)
             if p == 0:
-                self.empty_squares = set_nth_bit(i, self.empty_squares, True)
+                empty_squares = set_nth_bit(i, empty_squares, True)
+        return ProbabilityHistory(
+            repetitions=repetitions,
+            probabilities=probabilities,
+            full_squares=full_squares,
+            empty_squares=empty_squares,
+        )
 
-    def _generate_accumulations(
-        self, repetitions: int = 1000, use_cache: bool = False
-    ) -> None:
+    def _generate_accumulations(self, repetitions: int = 1000) -> None:
         """Samples the state and generates the accumulated
         probabilities of each square, empty_squares, and full_squares.
         """
+        if self.move_history_probabilities_cache[-1].repetitions >= repetitions:
+            return
 
-        if use_cache and self.move_history:
+        if len(self.move_history) > 0:
             last_move = self.move_history[-1]
-            if self.move_history_probabilities_cache[-1][0] >= repetitions:
-                self.probabilities = self.move_history_probabilities_cache[-1][1].copy()
-                self._set_full_empty_squares_from_probability()
-                return
             previous_move_in_cache = (
-                len(self.move_history) > 1
-                and self.move_history_probabilities_cache[-2][0] >= repetitions
+                self.move_history_probabilities_cache[-2].repetitions >= repetitions
```

Comment:
Won't this throw a KeyError if you get rid of the other conditions and it only has one entry?

Reply:
After this refactoring, there is a cache entry for the initial position. So move_history_probabilities_cache is always longer by 1 element than move_history.

Reply:
Optional: Is it worth documenting this behavior somewhere (maybe where we initialize move_history_probabilities_cache) so that it is clear what this structure is and how it works?

Reply:
I did add a comment about this invariant in `__init__`